lib/clibuddy/command.rb

```diff
@@ -0,0 +1,65 @@
+module CLIBuddy
+  class Command
+
+    attr_reader :name
+    attr_accessor :flow, :definition, :usage
+
+    def initialize(name)
+      @name = name
+    end
+
+    def required_args
+      return [] if definition.arguments.empty?
+      definition.arguments.select { |a| !a.optional? }
+    end
+
+    def optional_args
+      return [] if definition.arguments.empty?
+      definition.arguments.select { |a| a.optional? }
+    end
+
+    # TODO This should probably be done in the formatter, or somewhere else (maybe>)
+    # when user types `--help` then we show the full usage text plus usage string,
+    # if they typo a command/used unrecognized arguments we show the short usage test plus usage string.
+    def full_usage
+      u = usage[:full].join("\n")
+      u += "\n\nUsage: #{name} "
+      u += required_args.map { |a| a.name }.join(" ")
+      u += " [options...]\n    "
+      oa = optional_args.map { |a| a.short_name + ", " + a.long_name.ljust(30) + a.description.join(" ") }
+      u += oa.join("\n    ")
+    end
+
+  end
+
+  CommandDefinition = Struct.new(:arguments)
+
+  class CommandDefinitionArg
+    attr_reader :name, :param, :description
+
+    def initialize(name, param, description)
+      @name = name
+      @param = param
+      @description = description
+    end
+
+    def optional?
+      name.start_with?("--")
+    end
+
+    def short_name
+      if optional?
+        i = name.index("*")
```

Comment:
If the `*` is not found a `nil` is returned. This could throw an error. I think it would be fair to assume the first character after the `--`.

Reply:
`i = name.index("*") or name[2]` this assumes that the double-dash always used for full names which feels pretty safe if you defining this only for testing the Linux platform.